modules/core/packer/roles/nomad/files/configure.sh

```diff
@@ -202,13 +204,97 @@ EOF
   chown "${user}:${user}" "${config_dir}/acl.hcl"
 }
 
+function generate_docker_config {
+  local readonly consul_prefix="${1}"
+  local readonly config_dir="${2}"
+  local readonly user="${3}"
+  local readonly consul_template_config="${4}"
+  local readonly docker_auth_path="${5}"
+
+    local docker_config=$(cat <<EOF
+client {
+  options {
+    "docker.auth.config" = "${docker_auth_path}"
+  }
+}
+EOF
+)
+
+  log_info "Writing Docker configuration to ${config_dir}/docker.hcl"
+  echo "${docker_config}" > "${config_dir}/docker.hcl"
+  chown "${user}:${user}" "${config_dir}/docker.hcl"
+
+  local vault_path
+  vault_path=$(consul_kv "${consul_prefix}docker-auth/path")
+
+  local docker_template=$(cat <<EOF
+template {
+  destination = "${docker_auth_path}"
+  create_dest_dirs = true
+
+  # consul-template does not deal with ownership properly
+  # See https://github.com/hashicorp/consul-template/issues/1061
+  command = "bash -c 'chown ${user}:${user} ${docker_auth_path}'"
+
+  perms = 0600
+  error_on_missing_key = true
+
+  # The goal is to produce something like
+  # {
+  #     "auths": {
+  #         "registry.a.b": {
+  #           "auth": "aaaaa="
+  #         },
+  #         "foo.bar.xyz": {
+  #           "auth": "bbbb="
+  #         }
+  #     }
+  # }
+  #
+  # But two things make it difficult:
+  # 1. JSON doesn't do dangling commas
+  # 2. Go Template has really basic logic handling
+  #
+  # Thus this template is more complicated than needed.
+  contents = <<EOH
+{{- define "keys" -}}
+  {{- with secret "${vault_path}" }}
+      {{- range \$key, \$value := .Data -}}
+        {{ \$key }}{{ " " }}
+      {{- end -}}
+  {{- end -}}
+{{- end -}}
+{{- \$keys := (executeTemplate "keys" | trimSpace | split " ") -}}
+{
+    "auths": {
+    {{- with secret "${vault_path}" -}}
+    {{ \$auths := .Data }}
+      {{- range \$i, \$key := \$keys }}
+        {{- if \$i }},{{ end }}
+        "{{ \$key }}": {
+          "auth": "{{ index \$auths \$key }}"
+        }
+      {{- end }}
+    {{- end }}
+    }
+}
+EOH
+}
+EOF
+)
```

Comment:
Let me guess; Go template doesn't support index + key/value pair iteration directly, i.e. `range` only supports either Array style (index with value), or Object style (key with value), which is why you had to separately define a template to flatten key-value Object into just a mere Array. Ref: https://golang.org/pkg/text/template/#hdr-Variables

After which, scope the flattened Array within the original Object again, then now you have access to range of Array to give you the index + key, and value is derived by accessing the Object via key index.

The comma part is done prepended; a trick on indexing where at index 0, it doesn't execute. Ref: https://stackoverflow.com/questions/42662525/go-template-remove-the-last-comma-in-range-loop

Is the explanation correct?

Reply:
Yes.

Also, can't do index based enumeration with a map. No `map` function for a map. **ARGH**